### ISSUE TITLE
scrape (histograms): Automatically reduce resolution rather than fail scrape

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#12864]
### Proposal

Currently, we fail the entire scrape if the configured bucket limit is exceeded in a native histogram (scrape config option `native_histogram_bucket_limit`). It would be better if we simply reduced the resolution (in other words: decreased the schema) sufficiently to be within the `native_histogram_bucket_limit`. This is computationally easy (just adding up neighboring buckets).

Once we have done this, it might also be useful to add another scrape config option like `native_histogram_max_schema` to limit the resolution overall (and automatically lower the resolution of scraped histograms accordingly). This avoids needlessly storing histograms at a higher resolution than needed (but gives other scrapers still the option to scrape the histogram at the full resolution provided by the target).

This is particularly relevant with OTel exponential histograms, as they follow the strategy to start with a very high resolution and only reduce it once the bucket limit is reached. While this approach minimizes the config required by the user, it creates a lot of "resolution change noise". On the backend side, there is usually a good idea of a maximum useful resolution (e.g. because histograms get aggregated over time or across labels anyway, thereby using the lowest common resolution). Paying the storage resources to store the frequent resolution changes as well as the short periods of higher resolution should be avoided.

(For reference, Mimir has a similar [feature request](https://github.com/grafana/mimir/issues/5367), but I think this is definitely useful here in upstream Prometheus.)

**Comment by beorn7** [comment]
This should also apply to the remote write receiver (despite the "scrape" prefix in the issue title). AFAIK, we don't even have a bucket limit yet in the remote write receiver.

**Comment by krajorama** [comment]
@gouthamve @jesusvazquez This is issue is relevant for the OTLP endpoint as well. 

I think it would be great if Prometheus provided an interface to downscale a histogram to keep the desired schema and bucket number limit. Something like 

```
// github.com/prometheus/prometheus/model/histogram
// Shrink reduces the resolution of the Histogram to fit the given constraints. Use math.MaxInt32 for no limit.
func (h *Histogram) Shrink(maxSchema int32, maxBuckets int32) (h *Histogram)
```

We could potentially reuse the interface in the Mimir distributor (before replication) and also in the OTLP endpoint.

Open question: whether in OTLP we should do the shrink during/before converting to native histogram. This can probably be postponed to an optimization issue.

**Comment by beorn7** [comment]
I think such a method makes sense (I would use another name, but we can discuss that in a PR).

**Comment by fatsheep9146** [comment]
how about we convert the function `mergeToSchema`  into `Shrink` like @krajorama suggested into a more generic function like `compactBuckets` to cover the histogram and float histogram?

https://github.com/prometheus/prometheus/blob/ffa74eb12db3d46c98033cea42cfdb347e7d1e75/model/histogram/float_histogram.go#L916

https://github.com/prometheus/prometheus/blob/ffa74eb12db3d46c98033cea42cfdb347e7d1e75/model/histogram/generic.go#L130

@beorn7 

If this way is ok, I'd love to help with it

**Comment by beorn7** [comment]
That might make sense. Depends on how the implementation works out.

@krajorama might already be working on this. Maybe you two sync with each other on CNCF Slack?

**Comment by krajorama** [comment]
@fatsheep9146 Currently I'm working on other things, not sure when I get to this, so please go ahead. Reusing mergeToSchema sounds like a good idea, probably no need to rename it , it could be called from the shrink (or whatever) function!? I'm a little unsure about generics due to the big difference between how the bucket counts are actually stored in histogram vs float histogram, i.e. delta encoding integers vs absolute floats. I do a little experiment to benchmark the new functions with and without generics.

In either case, I'd advise to start writing up the test cases so we can agree on semantics , then do the simplest implementation and then we can look at performance as last step.

**Comment by fatsheep9146** [comment]
> @fatsheep9146 Currently I'm working on other things, not sure when I get to this, so please go ahead. Reusing mergeToSchema sounds like a good idea, probably no need to rename it , it could be called from the shrink (or whatever) function!? I'm a little unsure about generics due to the big difference between how the bucket counts are actually stored in histogram vs float histogram, i.e. delta encoding integers vs absolute floats. I do a little experiment to benchmark the new functions with and without generics.
> 
> In either case, I'd advise to start writing up the test cases so we can agree on semantics , then do the simplest implementation and then we can look at performance as last step.

Thanks! I will start writing the test cases first. @krajorama

**Comment by fatsheep9146** [comment]
#13001

In this pr, I already wrote a first version of generic `shrink` function and wrote two unit tests separately for `FloatHistogram` and `Histogram`, could you help check if we can agree with the behavior of the test and shrink function itself?

@krajorama @beorn7

**Comment by fatsheep9146** [comment]
After #13001 is merged, I think the issue still has two TODOs:
- [ ] when `native_histogram_bucket_limit` is hitted, reduce the resolution until it is beyond the limit.
- [ ] add new `native_histogram_max_schema` config 

@beorn7

**Comment by beorn7** [comment]
Sounds good.
